File: m02_ros2_communication/m02_ros2_with_py/m02_ros2_with_py/add_two_nums_cli.py
# !/usr/bin/env python3

# Sys is needed to read args passed with the executable
import sys

# ------------------ ROS2 Dependencies ----------------------------------------
import rclpy
from rclpy.node import Node

# ------------------ ROS2 Message/Srv Definition -----------------------------
from example_interfaces.srv import AddTwoInts

# ------------------ Client Node Class ----------------------------------------
class SimpleAdderCli(Node):
    """
    Simple class related with addition clients

    Attributes
    ---
    cli : Client object
        Client that will make request (two nums) using the '/add_two_ints'
    req : Request object
        Request that contains two double nums

    Methods
    ---
    cli_request(a,b)
        Make request to obtain sum of 'a' and 'b'
    """
    def __init__(self):
        """
        Constructor that initialize the client node, sets the client for the
        addition of two nums and wait until service is ready.
        """
        super().__init__("add_two_nums_cli")
        self.cli = self.create_client(AddTwoInts, 'add_two_ints')
        while not self.cli.wait_for_service(timeout_sec = 1.0):
            self.get_logger().info('Waiting for server...')
        self.req = AddTwoInts.Request()

    def cli_request(self, a, b):
        """
        Sends the client request for the sum of two numbers and return the
        result obtained.

        Params
        ---
        a : float
            First number to sum
        b : float
            Second number to sum

        Returns
        ---
        result : float
            Sum of the two numbers
        """
        self.req.a = a
        self.req.b = b
        self.future = self.cli.call_async(self.req)
        rclpy.spin_until_future_comple(self, self.future)
        return self.future.result()

# ---------------------------- Implementation --------------------------------
def main():
    # Initialize ROS Client Library for Python
    rclpy.init()

    # Intance adder client object
    cli_adder = SimpleAdderCli()
    
    # Get response by waiting to request to be received
    response = cli_adder.send_request(int(sys.argv[1]), int(sys.argv[2]))

    # Log result obtainer
    cli_adder.get_logger().info('Result of the addition is: %d', response.sum)

    # Destroyed of class and shutdown of resources used
    cli_adder.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: m02_ros2_communication/m02_ros2_with_cpp/src/add_two_nums_cli.cpp
/**
 * File: add_two_nums_srv.cpp
 * 
 * Description: Simple client for adding two nums with the rclcpp 
 * implementation for ROS2
*/

// --------------------------- CPP Standard Libraries -------------------------

#include <chrono>  // For flexible collecction of types that tracks time with
                   // varying types of precision.

#include <cstdlib> // General purpose function

#include <memory>  // General utilities to manage dynamic memory

// --------------------------- rclcpp Headers ---------------------------------

#include "rclcpp/rclcpp.hpp"   // Proper include for using rclcpp

#include "example_interfaces/srv/add_two_ints.hpp"  // AddTwoInts service

// --------------------------- Implementation of namespaces -------------------
using namespace std::chrono_literals;

// --------------------------- MAIN PROGRAM -----------------------------------
int main(int argc, char **argv)
{
  // ROS CLient Library for C++ initialization
  rclcpp::init(argc, argv);

  // Check that are enough arguments "node + operands (2)"
  if (argc != 3) 
  {
    RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "usage: add_two_ints_client X Y");
    return 1;
  }

  // Node init with name "add_two_ints_client" by defining a shared pointer
  std::shared_ptr<rclcpp::Node> node = 
    rclcpp::Node::make_shared("add_two_ints_client");

  // Instance a service client objet that sends request of AddTwoInts by the
  // service 'add_two_ints'.
  rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedPtr client =
    node->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");

  // Instance a request object and assign the two numbers as summands, atoll is
  // needed to make the conversion from C-String to long long int.
  auto request = 
    std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
  request->a = atoll(argv[1]);
  request->b = atoll(argv[2]);

  // Loop for waiting to the server to be online (logs message each second)
  while (!client->wait_for_service(1s)) 
  {
    // Exits the program if something goes wrong
    if (!rclcpp::ok()) 
    {
      RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), 
                   "Interrupted while waiting for the service. Exiting.");
      return 0;
    }
    RCLCPP_INFO(rclcpp::get_logger("rclcpp"), 
                "service not available, waiting again...");
  }

  // After connection is made, receive the asynchronous request
  auto result = client->async_send_request(request);

  // Wait for the result and check if the value is correctly received
  if (rclcpp::spin_until_future_complete(node, result) ==
    rclcpp::FutureReturnCode::SUCCESS)
  {
    RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Sum: %ld", result.get()->sum);
  } 
  else 
  {
    RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), 
                 "Failed to call service add_two_ints");
  }

  // Close after printing result
  rclcpp::shutdown();
  return 0;
}